A Starting Point

AttonRisk opened the pull request:
This is my first time using git in any genuine capacity. The code below isn't perfect, but most of this can be updated regardless.

AttonRisk committed:
A Starting Point

A starting point

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+------------------------------------------------------------------------------------------------
+]]
+
+f,d,a = 112,219,244 
```
Please expand these variables to full words.
What do they represent?
If they are constants name them with full upper case and underscores e.g. `CONSTANT_VALUE = 123`

AttonRisk replied:
Was replaced by the following since there was no reason to have those values outside of the function itself. 
`function getClientInfo ()
	local GIRED,GIGREEN,GIBLUE = 112,219,244 
	local clientIp, clientName, clientSerial = getPlayerIP(client), getPlayerName(client), getPlayerSerial(client)
	outputChatBox(tostring(na),client,GIRED,GIGREEN,GIBLUE)
	outputChatBox(tostring(se),client,GIRED,GIGREEN,GIBLUE) 
	outputChatBox(tostring(ip),client,GIRED,GIGREEN,GIBLUE) 
end`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+	if plr then
+		local ip = getPlayerIP(plr) 
+		local na = getPlayerName(plr) 
+		local se = getPlayerSerial(plr) 
```
Please expand these variable names to full words eg `playerName`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+	local se = getPlayerSerial(plr) 
+	outputChatBox(tostring(na),plr,f,d,a) 
+	outputChatBox(tostring(se),plr,f,d,a) 
+	outputChatBox(tostring(ip),plr,f,d,a) 
```
Please expand these variables to full words. But in reality lines 42-44 can be replaced by `local ip, playerName, playerSerial = getInfoData(player)` 

I'd also recommend changing `getInfoData` to something like `getPlayerInfo`.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.RespawnPlayer", true) 
+addEventHandler("aCore.RespawnPlayer", getRootElement(), respawn) 
+
+function GETINFO (plr)
```
Please use camel case for function names `lowerUpperUpper (` for private and `UpperUpperUpper (`for public.

Replace `GETINFO` with `outputPlayerInfo` since this outputs to the chatbox.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.getPlayerInfo", true) 
+addEventHandler("aCore.getPlayerInfo", getRootElement(), GETINFO) 
+
+function FIXCARf (plr)
```
Please use camel case for function names

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.ssEngine", true) 
+addEventHandler("aCore.ssEngine", getRootElement(), toggleEngine) 
+
+function FullHealf ()
```
Please change to `FullHeal` or `fullHealth`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.healPlayer", true) 
+addEventHandler("aCore.healPlayer", getRootElement(), FullHealf) 
+
+function blowup ()
```
Please change function name to camelcase

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.blowUpPlayer", true) 
+addEventHandler("aCore.blowUpPlayer", getRootElement(), blowup) 
+
+function Cloakf(plr)
```
What does the f symbolise here? I would suggest naming the function `togglePlayerCloaking`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEvent("aCore.cloak", true) 
+addEventHandler("aCore.cloak", getRootElement(), Cloakf) 
+
+function Cloakcf(plr)
```
What does cf symbolise here? I would suggest renaming the function to `togglePlayerVehicleCloaking`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEventHandler("aCore.cloak", getRootElement(), Cloakf) 
+
+function Cloakcf(plr)
+local DM = getPedOccupiedVehicle(plr)
```
Please change `DM` to `playerVehicle`. Also fix indentation on this line.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+local DM = getPedOccupiedVehicle(plr)
+	if (isPedInVehicle(plr)) then 
+		if (not DM) then
+	    return 
```
Is this required? In what scenario will `getPedOccupiedVehicle` not return a vehicle while `isPedInVehicle` returns true?
If there isn't a scenario I would suggest changing this to just return (possibly return message to user) and change following else to the start of an if.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEventHandler("aCore.cloakCar", getRootElement(), Cloakcf) 
+
+
+function DMGf (plr)
```
Please change this function name to `toggleVehicleDamageProof` or similar.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+
+
+function DMGf (plr)
+local DM = getPedOccupiedVehicle(plr) 
```
Please rename this variable to `playerVehicle` and fix indentation.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEventHandler("aCore.damageProof", getRootElement(), DMGf) 
+
+function BLOWf (plr)
+local BM = getPedOccupiedVehicle (plr) 
```
Please change this function to camelCase and rename to `blowVehicle` or something similar.
Please rename `BM` to `playerVehicle`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+addEventHandler("aCore.blowUpCar", getRootElement(), BLOWf) 
+
+
+function Gmf (thePlayer)
```
Please rename this function to `toggleGodMode`

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+	if (isPedInVehicle(plr)) then 
+		if (not DM) then
+		return 
+		end 
```
Same as previous function:
"Is this required? In what scenario will `getPedOccupiedVehicle` not return a vehicle while `isPedInVehicle` returns true?
If there isn't a scenario I would suggest changing this to just return (possibly return message to user) and change following else to the start of an if."

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+function killaPlayer (plr)
+	if plr then
+		if killPed(plr) then
+			outputChatBox("You are dead not big surprise!", plr ,0 ,255,0) 
```
Please fix this sentence: "You are dead not _a_ big surprise!"

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+			outputChatBox("You are dead not big surprise!", plr ,0 ,255,0) 
+		else
+			-- Place Holder
+		end
```
In what scenario does killPed return false? What action do you want to do if it does? Notify user? Return error? Please fill this in.

AttonRisk replied:
Removed

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -0,0 +1,204 @@
+function fullKit (plr)
+	for a,b in pairs(ids) do
+		setPedArmor ( plr, 100 ) 
+		giveWeapon(plr, b["id"]) 
```
What do these ids represent? Please change the variable ids to weaponIds or similar to show what it represents.
Additionally please change the ids to represent allowed weapons. The server going forward will likely ban ranged weapons so stick with Hand/Melee/Special 1/Gits/Special 2. https://wiki.multitheftauto.com/wiki/Weapons
Suggestions:
Change ids to:
`local weaponIds = {1, 5, 42,46} --BrassKnuckle, Bat, Fire Extinguisher, Parachute`

Move the armour out of the loop since it doesn't need to be there.

Make the loop easier to workout what's going on in one go
Change loop to:

``` lua
for index, id in pairs(weaponIds) do
    giveWeapon(plr, id)
end
```

I know `index` sucks vs the traditional `i`, but for a non lua developer it's obvious what pairs does here if we give the full name. If we were using a 2d table then we'd probably use `key` here or whatever the key represents.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/client.lua:
```diff
@@ -0,0 +1,320 @@
+--	stopb = guiCreateButton(420, 375, 101, 51, "Stop", false, Agui)
+	TBDEV = guiCreateEdit(10, 335, 113, 21, " MP3 URL", false, Agui)
+	guiEditSetMaxLength(TBDEV, 32767)
+--[[===============================================================================================================================]]--	
```
What resolutions have these values been tested in? Some weird stuff can happen with different resolutions and aspect ratios, have a quick test with a few different ones.

Please use consistent naming for your ui elements, my suggestion would be `btnClose` etc. Make sure to use full words and no shortenings except for the type of element (https://gist.github.com/hosuaby/fe180dc02238c16e6be2 for other prefixes)

AttonRisk replied:
1024x720 and up have all been tested and proven to work, anything smaller caries the death penalty.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/client.lua:
```diff
@@ -0,0 +1,320 @@
+addEventHandler("onClientVehicleExit", getRootElement() , stickyf)
+addEventHandler("onClientVehicleExit", getLocalPlayer() , killSticky)
+addCommandHandler("killsticky",killSticky)
+]]--
```
Orphaned code should not be committed to source control

OpenGroveProject commented on resources/[freeroam]/OGPcm/Code/client.lua:
```diff
@@ -0,0 +1,63 @@
+function spawnRamp(cn,rampID)
+
+	if rampID then
+		wipeRamp()
```
I think you can get a much better user experience if you change this around a little.
Currently this will wipe the old ramp every 30 seconds after the resource is started, this could be 1 second after the `spawnRamp` command is called.

Try:
```lua
function wipeRamp(ramp)
    if isElement(ramp) then
        destroyElement(ramp)
    end
end

function spawnRamp(cmd, rampId)
    if rampId then
        wipeRamp(oldRamp)
.....
        setTimer(function()
            wipeRamp(oldRamp)
        end,1000*30,1)
    else
        outputChatBox("Use a vaild ramp!",255,0,0) --Spell check your chat box please
    end
end
addCommandHandler("ramp", spawnRamp)
```
Doing it this way means that the user will only have one ramp at a time, but they get to keep each ramp for 30 seconds if they don't create a new one rather than 0-30 seconds if we do a timer from resource start.

OpenGroveProject replied:
`.....` represents lines 48 -> 58 in the original file.

OpenGroveProject reviewed (changes requested):
Thanks for your contribution!
It's looking pretty good but a few things I'd like to see before pushing to the server.

`OGPcheats/Code/backend.lua` has a lot of suggestions for variable and function renaming, this is the kind of thing that really helps you in 3 months when you need to make a bug fix or other developers understand your code when they're working on it, please ensure this is carried out in the other files too.

In your server side code you should use `client` instead of plr where you are attempting to pass `localplayer` from the client code. The global variable `client` will always be the exact player that triggered your server side event through `triggerServerEvent` see https://wiki.multitheftauto.com/wiki/TriggerServerEvent for usage and https://wiki.multitheftauto.com/wiki/Script_security for the why.

One last thing I'd like to see before approving is some big `--TODO - <ADD USER PERMISSIONS>` comments over functions that look like they could be disruptive if everyone was allowed to use them, especially the kill another player functions.

In the future I'd like to see each resource in a separate pull request, but happy to keep this one open until approved.

Please ask any questions you have and I'll answer as soon as I see them.

AttonRisk committed:
Fixed Code

Attached is the code with the adjustments and improvement required. If I have missed anything, please inform me.

AttonRisk commented:
A large number of adjustments have been made as per feedback in the latest commit

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+		if (getVehicleEngineState(clientVeh)) then
+			setVehicleEngineState(clientVeh, false) 
+			outputChatBox("Engine OFF!", client, 255, 0, 0) 
+			local engineStatus = false 
```
Unused variable, please delete.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+			setVehicleEngineState(clientVeh, false) 
+			outputChatBox("Engine OFF!", client, 255, 0, 0) 
+			local engineStatus = false 
+		else
+			if not getVehicleEngineState(clientVeh) then
```
This if will always evaluate to true since we're in the else of the opposite, delete the if.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+			if not getVehicleEngineState(clientVeh) then
+				setVehicleEngineState(clientVeh, true) 
+				outputChatBox("Engine ON!",client, 0, 255, 0) 
+				local engineStatus = true 
```
Unused variable, please delete.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+addEventHandler("OGP.clientExplodePlayer", getRootElement(), clientExplode) 
+
+function cloakClient()
+    if getElementAlpha(client) == 255 then
```
What about if their alpha is somewhere between 255 and 0? Do we want to cloak them if over 0?

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+	if (isPedInVehicle(client)) then 
+		if (not clientCar) then
 	    return 
     end 
```
These two if's evaluate the same thing, delete one.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+	local clientCar = getPedOccupiedVehicle(client) 
+	if (isPedInVehicle(client)) then 
+		if getPedOccupiedVehicleSeat(client) == 0 then
+			if (not clientCar) then
+				return 
+			end 
```
This if is redundant, delete it. 
In fact it will never evaluate to true so will never be used.

OpenGroveProject commented on resources/[freeroam]/OGPcheats/Code/backend.lua:
```diff
@@ -2,203 +2,193 @@ hold =[[
+function killClient ()
+	if client then
+		if killPed(client) then
+			outputChatBox("You are dead not big surprise!",client,0,255,0) 
```
Still need to proof read those chat boxes! 
`"You are dead not _a_ big surprise!"`

OpenGroveProject commented on resources/[freeroam]/OGPcm/Code/client.lua:
```diff
@@ -31,33 +31,36 @@ function getPositionInfrontOfElement ( element , meters )
+				wipeRamp()
+				if tonumber(rampID) then
+					rampID = tonumber(rampID)
+					if not (rampID > 8) then
+						if not (1 > rampID) then
```
You might want to change these to `if rampID > 0 and rampID <= #rampTable then` It's more readable, and if we add more ramps later this line wont need to change. Feel free to use `rampID >= 1` if you want to really push home that you're evaluating the id is in the ramp table.

_Below is long winded and not necessary, but useful information if you wish to improve as a software developer._

Alternatively it might be more performant (if we end up with a large enough ramp table) to evaluate the ramp existing in the table instead.

```lua
--line 48: wipeRamp()
local rampObjectID = rampTable[tonumber(rampID)] --Get the object id for the ramp id we're passed.
 --If tonumber failed (returns 0) or rampTable[rampID] doesn't exist, rampObjectID will equal nil, which evaluates to false in an if
if rampObjectID then
    --do ramp logic
end
```
The above is better since we only convert to number once vs twice in your code:
```lua
---line 48 wipeRamp()
if tonumber(rampID) then
    rampID = tonumber(rampID)
```
We only have one `if` to evaluate vs 3 in yours.
And vs my first suggestion `if rampID > 0 and rampID <= #rampTable then` it will be more performant since `#rampTable` is the equivalent of
```lua
local count = 0;
for index, value in ipairs(rampTable) do
    count++
end
return count
```
This is [O(n)](https://en.wikipedia.org/wiki/Big_O_notation) in perfomance but our look up is [O(1)](https://en.wikipedia.org/wiki/Big_O_notation), i.e. it will be the exact same performance whether we have 1 item or 1000, where as `#rampTable` will take 1 loop for 1 item or 1000 loops for 1000 items.



Now after all of that, it actually doesn't matter in this use case since we're probably not ever going to have that many items in our ramp table, but it's worth getting in the habit of thinking like this while you're coding.

OpenGroveProject reviewed (approved):
So I'm happy enough with the changes, there are still some unused variables and some extra if's that we don't need, I also wrote a comment about evaluating performance and big O notation for the rampID logic, but that all got wiped by a browser crash.

I might come back and comment on them for your learning if you'd like, but for now I'll push this to the server.

OpenGroveProject commented:
> So I'm happy enough with the changes, there are still some unused variables and some extra if's that we don't need, I also wrote a comment about evaluating performance and big O notation for the rampID logic, but that all got wiped by a browser crash.
> 
> I might come back and comment on them for your learning if you'd like, but for now I'll push this to the server.

OK so github is better than me, the comments are here! Feel free to look at them and add the changes to this pull request, I'll leave it open for a bit if you'd like to do them.
